File: server-mh/src/Match/Handler/Actions/GameAction.ts
import { ActionType, PlayerColor } from "../Enums";
import { MatchContext } from "../Models/MatchContex";
import { Piece } from "../Models/Piece";
import { ActionResult } from "./ActionResult";
import { GameActionData } from "./Datas";

export class GameAction {
    public actionType: ActionType = ActionType.MoveAction;

    public playerColor: PlayerColor = PlayerColor.Blue;

    public pieceIndex: number = 0;

    public path: number[] = [];

    public result: ActionResult | undefined = undefined;

    public Apply(context: MatchContext): void {
        switch (this.actionType) {
            case ActionType.MoveAction:
                this.ApplyMove(context);
                break;

            case ActionType.SpawnAction:
                this.ApplySpawn(context);
                break;

            case ActionType.ActivateSafeCellAction:
                this.ApplyActiveSafeCell(context);
                break;

            case ActionType.ActivatePenaltyCellAction:
                this.ApplyActivatePenaltyCell(context);
                break;
        }
    }
    private ApplyMove(context: MatchContext): void {
        const player =
            context.state.players[this.playerColor];

        const piece =
            player.pieces[this.pieceIndex];

        if (this.path.length > 0) {

            const destinationCell =
                context.state.board.cells[
                this.path[this.path.length - 1]
                ];

            piece.currentCell = destinationCell;
        }
        piece.pieceState.hasLeftStart = true;

        if (!this.result) {
            context.logger.info('result is null');
            return;
        }

        if (this.result.capturedEnemyColor != null && this.result.capturedEnemyIndex != null) {

            context.logger.info(`GameAction: ApplyMove: capturedEnemyPlayerColor=${this.result.capturedEnemyColor},capturedEnemyPieceIndex= ${this.result.capturedEnemyIndex}`);
            const capturedEnemy =
                context.state.players[this.result.capturedEnemyColor]
                    .pieces[this.result.capturedEnemyIndex];
            this.ResetPiece(capturedEnemy, context);
            context.state.turnState.hasReward = true;
        }

        if (this.result.enteredPenaltyCell)
            this.ResetPiece(piece, context);

        if (this.result.pieceFinish)
        {
            
            context.state.turnState.hasReward = true;
            piece.pieceState.finished = true;
        }

        if (this.result.playerFinish) {

            player.playerState.isFinished = true;

            context.state.winnerList.push(
                this.playerColor
            );
        }

        if (this.result.matchFinish)
            context.state.matchFinish = true;
    }
    private ApplySpawn(context: MatchContext): void {

        const player =
            context.state.players[this.playerColor];
        player.playerState.spawnedBefore = true;
        const piece =
            player.pieces[this.pieceIndex];

        const startCell =
            context.state.board
                .config
                .playerPath[player.color]
                .startHomeEntryCell;
        piece.currentCell =
            context.state.board.cells[startCell];
        context.logger.info(`GameAction: ApplySpawn: playerColor: ${this.playerColor}, pieceIndex: ${this.pieceIndex}, currentCell: ${piece.currentCell.index}`);
        piece.pieceState.spawned = true;
    }
    private ApplyActiveSafeCell(
        context: MatchContext
    ): void {

        const cell =
            context.state.board.cells[
            this.path[0]
            ];

        cell.isSafe = true;
    }
    private ApplyActivatePenaltyCell(
        context: MatchContext
    ): void {

        const cell =
            context.state.board.cells[
            this.path[0]
            ];

        cell.isPenalty = true;
    }
    public ToData(): GameActionData {

        const data = new GameActionData();

        data.Type = this.actionType;
        data.PlayerColor = this.playerColor;
        data.PieceIndex = this.pieceIndex;
        data.CellIndexes = [...this.path];

        if (this.result)
            data.Result = this.result.ToData();

        return data;
    }
    public FromData(
        data: GameActionData,
        context: MatchContext
    ): void {

        this.actionType = data.Type;
        this.playerColor = data.PlayerColor;
        this.pieceIndex = data.PieceIndex;
        this.path = [...data.CellIndexes];

        this.result = undefined;

        if (data.Result) {

            this.result = new ActionResult();

            this.result.FromData(
                data.Result,
                context
            );
        }
    }
    public ResetPiece(piece: Piece, context: MatchContext): void {
        context.logger.info(`Piece: Reset: playerColor: ${piece.player.color}, pieceIndex: ${piece.id}, currentCell: ${piece.currentCell.index}, initialCell: ${piece.initialCell.index}`);
        piece.currentCell = piece.initialCell;
        piece.pieceState.spawned = false;
        piece.pieceState.finished = false;
        piece.pieceState.hasLeftStart = false;


    }
    public Broadcast(context: MatchContext): void {
        context.broadcaster.NewAction(
            this
        );
        if (this.result) {
            if (this.result.capturedEnemyColor != null && this.result.capturedEnemyIndex != null) {
                const capturedEnemy =
                    context.state.players[this.result.capturedEnemyColor]
                        .pieces[this.result.capturedEnemyIndex];
                context.broadcaster.CapturePiece(capturedEnemy);
            }
            if(this.result.enteredPenaltyCell)
            {
                let piece = context.state.players[this.playerColor].pieces[this.pieceIndex];
                context.broadcaster.CapturePiece(piece);
            }
            if(this.result.playerFinish)
            {
                let player = context.state.players[this.playerColor];
                context.broadcaster.PlayerFinish(player);
            }
        }

    }

}

File: Client-unity/Ludo/Assets/Scripts/OfflineMatchHandler/GameAction.cs
using System.Collections.Generic;

namespace Ludo.Offline
{
    public class GameAction
    {
        public GameActionType ActionType = GameActionType.MoveAction;

        public PlayerColor PlayerColor = PlayerColor.Blue;

        public int PieceIndex = 0;

        public List<int> Path = new List<int>();

        public ActionResult Result;

        public void Apply(MatchContext context)
        {
            switch (ActionType)
            {
                case GameActionType.MoveAction:
                    ApplyMove(context);
                    break;

                case GameActionType.SpawnAction:
                    ApplySpawn(context);
                    break;

                case GameActionType.ActivateSafeCellAction:
                    ApplyActiveSafeCell(context);
                    break;

                case GameActionType.ActivatePenaltyCellAction:
                    ApplyActivatePenaltyCell(context);
                    break;
            }
        }

        private void ApplyMove(MatchContext context)
        {
            var player = context.State.Players[(int)PlayerColor];
            var piece = player.Pieces[PieceIndex];

            if (Path.Count > 0)
            {
                var destinationCell =
                    context.State.Board.Cells[
                        Path[Path.Count - 1]
                    ];

                piece.CurrentCell = destinationCell;
            }

            piece.PieceState.HasLeftStart = true;

            if (Result == null)
                return;

            if (Result.CapturedEnemy != null)
            {


                ResetPiece(Result.CapturedEnemy, context);

                context.State.TurnState.HasReward = true;
            }

            if (Result.EnteredPenaltyCell)
                ResetPiece(piece, context);

            if (Result.PieceFinish)
            {
                context.State.TurnState.HasReward = true;
                piece.PieceState.Finished = true;
            }

            if (Result.PlayerFinish)
            {
                player.PlayerState.IsFinished = true;

                var winnerList =
                    new List<PlayerColor>(
                        context.State.WinnerList ?? new PlayerColor[0]
                    );

                winnerList.Add(PlayerColor);

                context.State.WinnerList = winnerList.ToArray();
            }

            if (Result.MatchFinish)
                context.State.MatchFinish = true;
        }

        private void ApplySpawn(MatchContext context)
        {
            var player = context.State.Players[(int)PlayerColor];

            player.PlayerState.SpawnedBefore = true;

            var piece = player.Pieces[PieceIndex];

            var startCell =
                context.State.Board
                    .Config
                    .PlayerPath[player.Color]
                    .StartHomeEntryCell;

            piece.CurrentCell =
                context.State.Board.Cells[startCell];

            piece.PieceState.Spawned = true;
        }

        private void ApplyActiveSafeCell(MatchContext context)
        {
            var cell =
                context.State.Board.Cells[Path[0]];

            cell.IsSafe = true;

            context.State.Players[(int)PlayerColor]
                .PlayerState.HasSpecialSafeCell = false;
        }

        private void ApplyActivatePenaltyCell(MatchContext context)
        {
            var cell =
                context.State.Board.Cells[Path[0]];

            cell.IsPenalty = true;

            context.State.Players[(int)PlayerColor]
                .PlayerState.HasSpecialPenaltyCell = false;
        }

        public void ResetPiece(
            Piece piece,
            MatchContext context)
        {
            piece.CurrentCell = piece.InitialCell;

            piece.PieceState.Spawned = false;
            piece.PieceState.Finished = false;
            piece.PieceState.HasLeftStart = false;
        }

        public void EnqueueCommands(MatchContext context)
        {
            // 1. NewAction
            context.CommandHandler.Enqueue(
                new NewActionCommand(
                    new GameActionDto
                    {
                        Type = ActionType,
                        PlayerColor = PlayerColor,
                        PieceIndex = PieceIndex,
                        CellIndexes = new List<int>(Path)
                    }
                )
            );

            if (Result == null)
                return;

            // 2. Captured enemy
            if (Result.CapturedEnemy != null)
            {
                var piece = Result.CapturedEnemy;

                context.CommandHandler.Enqueue(
                    new CapturePieceCommand(
                        new PiecePositionDto
                        {
                            PlayerColor = piece.Player.Color,
                            PieceId = piece.Id,
                            CellIndex = piece.CurrentCell.Index
                        }
                    )
                );
            }

            // 3. Entered penalty
            if (Result.EnteredPenaltyCell)
            {
                var player =
                    context.State.Players[(int)PlayerColor];

                var piece =
                    player.Pieces[PieceIndex];

                context.CommandHandler.Enqueue(
                    new CapturePieceCommand(
                        new PiecePositionDto
                        {
                            PlayerColor = piece.Player.Color,
                            PieceId = piece.Id,
                            CellIndex = piece.CurrentCell.Index
                        }
                    )
                );
            }

            // 4. Player finish
            if (Result.PlayerFinish)
            {
                context.CommandHandler.Enqueue(
                    new PlayerFinishedCommand()
                );
            }
        }
    }
}